src/distributed-transaction/percolator.md

```diff
@@ -0,0 +1,153 @@
+
+* `c:lock`
+* `c:write`
+* `c:data`
+* `c:notify`
+* `c:ack_O`
+
+Percolator also relies on a service named *timestamp oracle*. The timestamp oracle can produce timestamps in strictly increasing order. All read and write operations need to apply for timestamps from the timestamp oracle, and a timestamp came from timestamp oracle will be regarded as the logical time when the read/write operation happens.
+
+Percolator is multi-version, and a data item's version is represented by the timestamp that the transaction was committed.
+
+For example, in such a state:
+
+| key | v:data | v:lock | v"write |
+|-----|--------|--------|---------|
+|k1   |14:"value2"<br/>12:<br/>10:"value1"|14:primary<br/>12:<br/>10:|14:<br/>12:data@10<br/>10:|
+
+It means that for the key `k1`, a value `"value1"` was committed at timestamp `12`. Then there is an uncommitted version whose value is `"value2"`, and it's uncommitted because there's a lock. You will understand why it looks like this after understanding how transactions work.
+
+The data of a certain row and a certain Percolator column is named a *cell*. The above example shows different versions of data of a single cell.
+
+The remaining columns, `c:notify` and `c:ack_O`, are used for Percolator's incremental processing. After a modification, we set `c:notify` column to mark a cell to be dirty. There can be some *observers* which can do some user-specified operations when they find data of their observed columns have changed. To find whether data is changed, they continuously scanning the `notify` columns to find dirty cells. `c:ack_O` is the "acknowledgment" column of the observer `O`, which is used to prevent a row from incorrectly being notified twice. It saves the timestamp of the observer's last execution.
+
+### Writing
+
+Percolator's transactions are committed by a 2PC algorithm. Its two phases are called `Prewrite` and `Commit`.
+
+In `Prewrite` phase:
+1. We get a timestamp from the timestamp oracle, and we call the timestamp the transaction's `start_ts`.
+2. For each row involved in the transaction, put a lock to `lock` column. At the same time, put the value to the `data` column with the timestamp `start_ts`. One of these locks will be chosen as the *primary* lock, and others are *secondary* locks. Each lock saves the transaction's `start_ts`. The secondary locks, in addition, save the primary lock's location.
+    * If we found that there's already a lock or newer version than `start_ts`, the current transaction will be rolled back.
+
+And then, in `Commit` phase:
+1. Get another timestamp, namely `commit_ts`.
+2. Remove the primary lock, and at the same time write a record to `write` column with the timestamp `commit_ts`, and its value records the transaction's `start_ts`.
+    * If the lock is missing, the commit should fail.
```

Comment:
which lock is missing? The primary lock to remove? Also if there is only one point, we may consider moving it to the previous line instead of having one bullet for it.